### PR TITLE
Riscv simd

### PR DESCRIPTION
**Pull request by vorosl**

### CODE REVIEW

File: src/jit/SimdRiscvInl.h [hunk 1]
```diff
@@ -0,0 +1,1266 @@
...
+    simdEmitOp(compiler, SimdOp::vand_vv, tmp, tmp, rt);
+    uint32_t opcode = SimdOp::vfirst_m | sljit_get_register_index(SLJIT_GP_REGISTER, tmpgp) << 7 | sljit_get_register_index(SLJIT_SIMD_REG_128, tmp) << 20;
+    sljit_emit_op_custom(compiler, &opcode, sizeof(uint32_t));
+    struct sljit_jump* jump = sljit_emit_cmp(compiler, SLJIT_SIG_LESS, tmpgp, 0, SLJIT_IMM, 0);
+    sljit_set_label(sljit_emit_jump(compiler, SLJIT_JUMP), label);
+    sljit_set_label(jump, sljit_emit_label(compiler));
+#endif
+}
+
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:348]: Single newline

File: src/jit/SimdRiscvInl.h [hunk 2]
```diff
@@ -0,0 +1,1266 @@
...
+    ASSERT(instr->next() != nullptr);
+
+    if (instr->info() & Instruction::kIsMergeCompare) {
+        return true;
+    }
+
+    return false;
+}
+
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:645]: Single newline.

File: src/jit/SimdRiscvInl.h [hunk 3]
```diff
@@ -0,0 +1,1266 @@
...
+    simdOperandToArg(compiler, operands + 1, args[1], srcType, instr->requiredReg(1));
+    simdOperandToArg(compiler, operands + 2, args[2], dstType, instr->requiredReg(2));
+
+    args[3].set(operands + 3);
+    sljit_s32 dst = GET_TARGET_REG(args[3].arg, instr->requiredReg(2));
+
+    /*if (!sljit_has_cpu_feature(SLJIT_HAS_AVX) && dst != args[2].arg) {
+        sljit_emit_simd_mov(compiler, SLJIT_SIMD_REG_128 | srcType, dst, args[2].arg, 0);
+        args[2].arg = dst;
+    }*/
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1113]: There are many commented out parts. Do you need them? We have no AVX on RISCV.

File: src/jit/SimdRiscvInl.h [hunk 4]
```diff
@@ -0,0 +1,1266 @@
...
+        }
+        args[1].argw %= div;
+    }
+
+    simdEmitTypedOp(compiler, type, op, dst, args[0].arg, isImm ? args[1].argw : args[1].arg, isImm ? SimdOp::rm_imm : SimdOp::rm_gpreg);
+
+    if (SLJIT_IS_MEM(args[2].arg)) {
+        sljit_emit_simd_mov(compiler, SLJIT_SIMD_STORE | SLJIT_SIMD_REG_128 | type, dst, args[2].arg, args[2].argw);
+    }
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1265]: Single newline.

File: src/jit/InstList.cpp [hunk 5]
```diff
@@ -422,11 +422,28 @@ void JITCompiler::dump()
...
+#else /* !SLJIT_SEPARATE_VECTOR_REGISTERS */
+#define PREFIX "F"
+#define SAVED_START SLJIT_FR(SLJIT_NUMBER_OF_SCRATCH_FLOAT_REGISTERS)
+#define SAVED_END SLJIT_FS0
+#endif /* !SLJIT_SEPARATE_VECTOR_REGISTERS */
+                            if ((variable.info & Instruction::TypeMask) == Instruction::V128Operand) {
+                                prefix = PREFIX;
+                                savedStart = SAVED_START;
+                                savedEnd = SAVED_END;
+                            } else {
```
Comment (zherczeg) [inline src/jit/InstList.cpp:438]: This does not look nice.

Keep the else part as default.

Then add an `#if` part for vector registers, which tests V128 and overwrites the `prefix` and other variables.

File: src/jit/RegisterAlloc.cpp [hunk 6]
```diff
@@ -20,6 +20,17 @@
 #include "jit/Compiler.h"
 #include <set>
 
+#if (defined SLJIT_SEPARATE_VECTOR_REGISTERS && SLJIT_SEPARATE_VECTOR_REGISTERS)
+#define SHORTIF(COND, VECTOR, FLOAT) \
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:24]: What about `VECTOR_SELECT`?

File: src/jit/RegisterAlloc.cpp [hunk 7]
```diff
@@ -87,26 +98,43 @@ class RegisterSet {
...
     uint8_t toCPUFloatReg(uint8_t reg)
     {
-        return m_floatSet.toCPUReg(reg);
+        return m_floatSet.toCPUReg(reg, SLJIT_FR0, SLJIT_FS0);
     }
 
+#if (defined SLJIT_SEPARATE_VECTOR_REGISTERS && SLJIT_SEPARATE_VECTOR_REGISTERS)
+    uint8_t toCPUVectorReg(uint8_t reg)
+    {
+        return m_vectorSet.toCPUReg(reg, SLJIT_VR1, SLJIT_VS0);
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:134]: SLJIT_VR1 is riscv only thing. Maybe you could introduce a macro for the first register, and use it everywhere. SLJIT_VR1 on riscv, SLJIT_VR0 everywhere else.

File: src/jit/RegisterAlloc.cpp [hunk 8]
```diff
@@ -799,9 +852,8 @@ void JITCompiler::allocateRegisters()
                         regs.floatReserve(reg + 1);
                     }
 #endif /* SLJIT_CONFIG_ARM_32 */
-
-                    regs.floatReserve(reg);
-                    instr->setRequiredReg(reuseTmpIndex, regs.toCPUFloatReg(reg));
+                    SHORTIF(((*nextType & Instruction::TypeMask) == Instruction::V128Operand), regs.vectorReserve(reg), regs.floatReserve(reg))
+                    SHORTIF(((*nextType & Instruction::TypeMask) == Instruction::V128Operand), instr->setRequiredReg(reuseTmpIndex, regs.toCPUVectorReg(reg)), instr->setRequiredReg(reuseTmpIndex, regs.toCPUFloatReg(reg)))
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:856]: You can use a `(a, b)` comma separated list for commands. This way you don't need to duplicate the macro (if).

E.g: `(regs.vectorReserve(reg), instr->setRequiredReg(reuseTmpIndex, regs.toCPUVectorReg(reg)))`

File: src/jit/RegisterAlloc.cpp [hunk 9]
```diff
@@ -916,6 +974,13 @@ void JITCompiler::allocateRegistersSimple()
         uint32_t tmpIndex = 0;
         uint32_t nextIntIndex = SLJIT_R0;
         uint32_t nextFloatIndex = SLJIT_FR0;
+#if (defined SLJIT_SEPARATE_VECTOR_REGISTERS && SLJIT_SEPARATE_VECTOR_REGISTERS)
+#if (defined SLJIT_CONFIG_RISCV && SLJIT_CONFIG_RISCV)
+        uint32_t nextVectorIndex = SLJIT_VR1;
+#else /* !SLJIT_CONFIG_RISCV */
+        uint32_t nextVectorIndex = SLJIT_VR0;
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:981]: If you have the macro for the first register, this will simplify.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/SimdRiscvInl.h [hunk 10]
```diff
@@ -0,0 +1,1251 @@
...
+    uint32_t avl = (uint32_t)1 << (4 - elem_size);
+
+    uint32_t opcode = VSETIVLI | (6 << 7) | (elem_size << 23) | (vlmul << 20) | (avl << 15);
+    sljit_emit_op_custom(compiler, &opcode, sizeof(uint32_t));
+}
+
+static void simdEmitOp(sljit_compiler* compiler, uint32_t opcode, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm, uint32_t optype = 0)
+{
+    rd = sljit_get_register_index((optype & SimdOp::rd_gpreg) ? SLJIT_GP_REGISTER : SLJIT_SIMD_REG_128, rd);
+    if (!(optype & SimdOp::rn_imm) && !(optype & SimdOp::rn_gpreg) && rn >= SLJIT_VR0) {
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:129]: Can rn be zero in this case? If not, it should be an assert. Similar cases below.

File: src/jit/SimdRiscvInl.h [hunk 11]
```diff
@@ -0,0 +1,1251 @@
...
+static void simdEmitCompare(sljit_compiler* compiler, sljit_s32 type, uint32_t opcode, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm, bool reverseMask = false)
+{
+    sljit_s32 mask = SLJIT_VR0;
+    simdEmitTypedOp(compiler, type, opcode, mask, rn, rm);
+    sljit_s32 tmp = SLJIT_TMP_DEST_VREG;
+    simdEmitOp(compiler, SimdOp::vmv_vi, tmp, 0, reverseMask ? (0x1F) : 0, SimdOp::rm_imm | SimdOp::rn_imm);
+    simdEmitOp(compiler, SimdOp::vmerge_vi, rd, tmp, reverseMask ? 0 : (0x1F), SimdOp::rm_imm);
+}
+
+static void simdEmitExtend(sljit_compiler* compiler, sljit_s32 type, bool low, bool s, sljit_s32 rd, sljit_s32 rn)
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:220]: s = signed? Could have a longer name. What is low?

File: src/jit/SimdRiscvInl.h [hunk 12]
```diff
@@ -0,0 +1,1251 @@
...
+#if (defined SLJIT_CONFIG_RISCV64 && SLJIT_CONFIG_RISCV64)
+        opC |= 0x53 | (0x79 << 25);
+#else
+        opC |= 0x53 | (0x69 << 25);
+#endif
+    }
+    sljit_emit_op_custom(compiler, &opC, sizeof(uint32_t));
+    simdEmitOp(compiler, SimdOp::vfadd_vf ^ SimdOp::vm, rd, rn, ftmp);
+}
+static void simdEmitFMinMax(sljit_compiler* compiler, sljit_s32 type, sljit_s32 opcode, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm)
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:297]: Newline before.

File: src/jit/SimdRiscvInl.h [hunk 13]
```diff
@@ -0,0 +1,1251 @@
...
+    simdEmitTypedOp(compiler, type, SimdOp::vfcvt_f_x_v, rd, tmp, 0, SimdOp::rm_imm);
+}
+
+static void simdEmitPMinMax(sljit_compiler* compiler, sljit_s32 type, bool min, sljit_s32 rd, sljit_s32 rn, sljit_s32 rm)
+{
+    sljit_s32 mask = SLJIT_VR0;
+    simdEmitTypedOp(compiler, type, SimdOp::vmflt_vv, mask, min ? rm : rn, min ? rn : rm);
+    simdEmitOp(compiler, SimdOp::vmerge_vv, rd, rn, rm);
+}
+static void simdEmitPopcnt(sljit_compiler* compiler, sljit_s32 type, sljit_s32 rd, sljit_s32 rn, sljit_s32 rt)
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:324]: Newline before.

File: src/jit/SimdRiscvInl.h [hunk 14]
```diff
@@ -0,0 +1,1251 @@
...
+        ASSERT_NOT_REACHED();
+        break;
+    }
+
+    if (SLJIT_IS_MEM(args[3].arg)) {
+        sljit_emit_simd_mov(compiler, SLJIT_SIMD_STORE | SLJIT_SIMD_REG_128 | dstType, dst, args[3].arg, args[3].argw);
+    }
+}
+
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1132]: Single newline.

File: src/jit/SimdRiscvInl.h [hunk 15]
```diff
@@ -0,0 +1,1251 @@
...
+    simdOperandToArg(compiler, operands + 2, args[2], dstType, instr->requiredReg(2));
+
+    args[3].set(operands + 3);
+    sljit_s32 dst = GET_TARGET_REG(args[3].arg, instr->requiredReg(2));
+
+    switch (instr->opcode()) {
+    case ByteCode::V128BitSelectOpcode:
+        break;
+    case ByteCode::I8X16RelaxedLaneSelectOpcode:
+    case ByteCode::I16X8RelaxedLaneSelectOpcode:
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1106]: Missing break.

File: src/jit/SimdRiscvInl.h [hunk 16]
```diff
@@ -0,0 +1,1251 @@
...
+
+static void emitShuffleSIMD(sljit_compiler* compiler, Instruction* instr)
+{
+}
+
+static void emitShiftSIMD(sljit_compiler* compiler, Instruction* instr)
+{
+    Operand* operands = instr->operands();
+    JITArg args[3];
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1155]: Single newline.

File: src/jit/SimdRiscvInl.h [hunk 17]
```diff
@@ -0,0 +1,1251 @@
...
+    uint32_t op = 0;
+    int div = 8;
+    int mask = 0x1F;
+
+    args[1].set(operands + 1);
+
+    const bool isImm = SLJIT_IS_IMM(args[1].arg);
+    sljit_s32 type = SLJIT_SIMD_ELEM_8;
+
+
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:1166]: Single newline.

**Review by zherczeg — CHANGES_REQUESTED** [review]

File: src/jit/SimdRiscvInl.h (hunk 11 above)
Comment (vorosl) [reply src/jit/SimdRiscvInl.h:220]: `low` is `true` if the operator is `ExtendLow`, and `false` if the operator is `ExtendHigh`

File: src/jit/SimdRiscvInl.h (hunk 10 above)
Comment (zherczeg) [reply src/jit/SimdRiscvInl.h:129]: `(rn == SLJIT_VR0) >= SLJIT_VR0`, so it is still true without an extra >=

File: src/jit/SimdRiscvInl.h [hunk 18]
```diff
@@ -0,0 +1,1249 @@
...
+    vssubu_vv = InstructionType::opivv | OPCODE(0x22),
+    vsub_vv = InstructionType::opivv | OPCODE(0x2),
+    vwmul_vv = InstructionType::opmvv | OPCODE(0x3B),
+    vxor_vi = InstructionType::opivi | OPCODE(0xB),
+    vxor_vv = InstructionType::opivv | OPCODE(0xB),
+    vzext_vf2 = InstructionType::opmvv | OPCODE(0x12) | (0x6 << 15),
+};
+
+enum OperandTypes : uint32_t {
+    rn_imm = 1 << 1,
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:109]: I think `rnImm` is the valid syntax in Walrus.

File: src/jit/SimdRiscvInl.h [hunk 19]
```diff
@@ -0,0 +1,1249 @@
...
+    rm_imm = 1 << 2,
+    rn_gpreg = 1 << 3,
```
Comment (zherczeg) [inline src/jit/SimdRiscvInl.h:111]: `rnIsGpr` sounds better to me. Maybe `rnIsImm` sounds better as well.

File: src/jit/RegisterAlloc.cpp [hunk 20]
```diff
@@ -87,26 +106,43 @@ class RegisterSet {
...
+        , m_vectorSet(SLJIT_NUMBER_OF_SCRATCH_VECTOR_REGISTERS, SLJIT_NUMBER_OF_SAVED_VECTOR_REGISTERS, false)
+#endif /* SLJIT_CONFIG_RISCV */
+#endif /* SLJIT_SEPARATE_VECTOR_REGISTERS */
     {
     }
 
     RegisterSet& integerSet() { return m_integerSet; }
     RegisterSet& floatSet() { return m_floatSet; }
-
+    // clang-format off
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:124]: Is the format is that bad? Maybe we could turn all one liners to three lines, where the { and } in different line.

File: src/jit/RegisterAlloc.cpp [hunk 21]
```diff
@@ -717,7 +778,7 @@ void JITCompiler::allocateRegisters()
 
             if (type & Instruction::FloatOperandMarker) {
                 if (resultVariable->reg1 != VariableList::kUnusedReg) {
-                    regs.floatReserve(resultVariable->reg1);
+                    VECTOR_SELECT((type == Instruction::V128Operand), regs.vectorReserve(resultVariable->reg1), regs.floatReserve(resultVariable->reg1))
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:781]: Do you need the () around `type ==`? For me it looks better without it.

**Review by zherczeg — CHANGES_REQUESTED** [review]
Only minor things remained.

File: src/jit/RegisterAlloc.cpp [hunk 22]
```diff
@@ -761,8 +829,9 @@ void JITCompiler::allocateRegisters()
                         regs.floatReserve(resultReg + 1);
                     }
 #endif /* SLJIT_CONFIG_ARM_32 */
-                    regs.floatReserve(resultReg);
-                    resultReg = regs.toCPUFloatReg(resultReg);
+                    VECTOR_SELECT(type == Instruction::V128Operand,
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:832]: I realized this form looks much better, than the single long line.

File: src/jit/RegisterAlloc.cpp [hunk 23]
```diff
@@ -87,26 +106,50 @@ class RegisterSet {
...
     {
     }
 
-    RegisterSet& integerSet() { return m_integerSet; }
-    RegisterSet& floatSet() { return m_floatSet; }
-
+    RegisterSet& integerSet()
+    {
+        return m_integerSet;
+    }
```
Comment (zherczeg) [inline src/jit/RegisterAlloc.cpp:125]: Add newline after each }

**Review by zherczeg — CHANGES_REQUESTED** [review]

**Review by zherczeg — APPROVED** [review]
LGTM

**Review by clover2123 — APPROVED** [review]
BTW is it possible to run our tests for RISC-V based JIT?

**Comment by zherczeg** [comment]
This patch focuses on simple implementations, some operations are still missing, so several tests will fail. We continue the work on fixing the missing tests.

**Pull request merged**